drivers/frequency/adf5902/adf5902.c

```diff
@@ -0,0 +1,1046 @@
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int32_t adf5902_write(struct adf5902_dev *dev, uint32_t data)
+{
+	uint8_t buff[ADF5902_BUFF_SIZE_BYTES];
+
+	buff[0] = data >> 24;
+	buff[1] = data >> 16;
+	buff[2] = data >> 8;
+	buff[3] = data;
+
+	return spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+}
+
+/**
+ * @brief Readback data from ADF5902.
+ * @param dev - The device structure.
+ * @param data - Store the read data.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int32_t adf5902_readback(struct adf5902_dev *dev, uint32_t *data)
+{
+	int32_t ret;
+	uint32_t i;
+	uint8_t buff[ADF5902_BUFF_SIZE_BYTES];
+
+	buff[0] = *data >> 24;
+	buff[1] = *data >> 16;
+	buff[2] = *data >> 8;
+	buff[3] = *data;
+
+	/* Write command */
+	ret = spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+	if (ret != SUCCESS)
+		return ret;
+
+	/* Send dummy value to get the data on DOUT */
+	buff[0] = ADF5902_SPI_DUMMY_DATA;
+	buff[1] = ADF5902_SPI_DUMMY_DATA;
+	buff[2] = ADF5902_SPI_DUMMY_DATA;
+	buff[3] = ADF5902_SPI_DUMMY_DATA;
+
+	ret = spi_write_and_read(dev->spi_desc, buff, ADF5902_BUFF_SIZE_BYTES);
+	if (ret != SUCCESS)
+		return ret;
+
+	/* Concatenate received data Bytes */
+	for (i = 0; i < ADF5902_BUFF_SIZE_BYTES; i++)
+		*data = (*data << 8) | buff[i];
+
+	return ret;
+}
+
+/**
+ * @brief Compute ADF4350 RF VCO frequency parameters.
+ * @param dev - The device structure.
+ * @return Returns 0 in case of success or negative error code.
+ */
+int32_t adf5902_vco_freq_param(struct adf5902_dev *dev)
+{
+
+	dev->f_pfd = (dev->ref_in * (1 + dev->ref_doubler_en) /
+		      (dev->ref_div_factor *(1 + dev->ref_div2_en)));
+
+	if (dev->f_pfd > ADF5902_MAX_FREQ_PFD)
+		return FAILURE;
+
+	dev->int_div = (uint16_t)(dev->rf_out / (dev->f_pfd * 2));
+
+	if (dev->int_div > ADF5902_MAX_INT_MSB_WORD)
+		return FAILURE;
+
+	dev->frac_msb = (uint16_t)((((float)(dev->rf_out) / (dev->f_pfd * 2)) -
+				    dev->int_div) * pow(2, 12));
+
+	if (dev->frac_msb > ADF5902_MAX_FRAC_MSB_WORD)
+		return FAILURE;
+
+	dev->frac_lsb = (uint16_t)((((((float)(dev->rf_out) / (dev->f_pfd * 2)) -
+				      dev->int_div) * pow(2, 12)) -
+				    dev->frac_msb) * pow(2, 13));
+
+	if (dev->frac_lsb > ADF5902_MAX_FRAC_LSB_WORD)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initializes the ADF5902.
+ * @param device - The device structure.
+ * @param init_param - The structure containing the device initial parameters.
+ * @return Returns 0 in case of success or negative error code.
+ */
+int32_t adf5902_init(struct adf5902_dev **device,
+		     struct adf5902_init_param *init_param)
+{
+	int32_t ret;
+	uint32_t i;
+	struct adf5902_dev *dev;
+
+	dev = (struct adf5902_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	/* GPIO Chip Enable */
+	ret = gpio_get(&dev->gpio_ce, init_param->gpio_ce_param);
+	if (ret != SUCCESS)
+		return ret;
```

Comment:
need to free dev on failure

Reply:
added unconditional jumps throughout the entire init function.